### PR TITLE
Azure blob storage integration

### LINKED ISSUE #171: Performing backups to Azure
Is it possible to backup Azure Blob Storage?

### PR DESCRIPTION
**Pull request by ilhanadiyaman**
Fixes #171

### CODE REVIEW

**Comment by adejanovski** [comment]
Hi @ilhanadiyaman,

thanks for bringing Azure storage support to Medusa!
Libcloud sadly isn't yet able to reliably and quickly upload very large files. For S3 we rely on awscli to handle large files (>100MB) uploads/downloads and for GCP we use gsutil.
I see that Azure has [its own cli](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-cli) which I think we'll need to use here as well. You could copy the S3 implementation for Azure as a starting point since gsutil has a funny way of getting its inputs...

I'll be happy to help you as much as I can.

Let me know what you think.

**Comment by ilhanadiyaman** [comment]
Hi @adejanovski,

I didn't know about libcloud reliability for large files. Thanks for letting me know, otherwise I was going to start using it on our cassandra cluster. :) I'll check what I can do with azure cli and let you know.

This work is still in progress. I'd like add some documentation and integration tips. Also, how can we add azure blob storage to CI/CD checks? I already add azure to the integration tests.

**Comment by arodrime** [comment]
Oh I'm late to the party: https://github.com/thelastpickle/cassandra-medusa/issues/171#issuecomment-715199497

@adejanovski is just too quick for me! :)

**Comment by ilhanadiyaman** [comment]
I didn't know either @adejanovski will look for the PR that fast. That's why I only let you @arodrime know last night :) But it's good to know that issue regarding large files. I'll check and let you guys know.

**Comment by ilhanadiyaman** [comment]
I just pushed the new commits regarding documentation and azcli implementation. During the testing, I increase the row numbers to be inserted to 10M in the first scenario so that the file size gets larger than 64MB. It took a lot of time for ccm to insert the rows but at least it didn't fail. If there are other way to test it, please let me know.

There is one thing to mention, S3 and GCS uses eTag header to return MD5 file hashes. While in Azure, eTag looks like a hex  which is calculated based on last modified time. When we reupload same files, these "eTag" hashes are changing which causes failure on validating the manifest. Therefore, I had to use another property which is called "CONTENT-MD5". 

Let me know what you think @adejanovski and @arodrime.

File: medusa/storage/azure_blobs_storage/azcli.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            cmd = [self._az_cli_path, "storage", "blob", "upload", "-f", str(src), "-c", bucket_name, "-n", dest]
+            objects.append(self.upload_file(cmd, dest, azcli_output))
+
+        return objects
+
+    def cp_download(self, *, src, bucket_name, dest, max_retries=5):
+        job_id = str(uuid.uuid4())
+        azcli_output = "/tmp/azcli_{0}.output".format(job_id)
+        objects = []
+        cmd = [self._az_cli_path, "storage", "blob", "download", "-f", str(dest), "-c", bucket_name, "-n", src]
```
Comment (adejanovski) [inline medusa/storage/azure_blobs_storage/azcli.py:69]:
```suggestion
        cmd = [self._az_cli_path, "storage", "blob", "download", "-f", os.path.join(str(dest), str(src).split("/")[-1]), "-c", bucket_name, "-n", src]
```

The typo fix from `downlod_blobs` to `download_blobs` along with making multi part upload threshold configurable made this code path testable during integration tests.
It fails consistently with the following error:
```
CLIInternalError: The command failed with an unexpected error. Here is the traceback:
[Errno 21] Is a directory: '/tmp/medusa-download-all-tables/system_auth/roles-5bc52802de2535edaeab188eecebb090/'
Traceback (most recent call last):
  File "/Users/adejanovski/projets/cassandra/thelastpickle/py36/lib/python3.6/site-packages/knack/cli.py", line 215, in invoke
    cmd_result = self.invocation.execute(args)
  File "/Users/adejanovski/projets/cassandra/thelastpickle/py36/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 654, in execute
    raise ex
  File "/Users/adejanovski/projets/cassandra/thelastpickle/py36/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 718, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
  File "/Users/adejanovski/projets/cassandra/thelastpickle/py36/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 709, in _run_job
    cmd_copy.exception_handler(ex)
  File "/Users/adejanovski/projets/cassandra/thelastpickle/py36/lib/python3.6/site-packages/azure/cli/command_modules/storage/__init__.py", line 341, in new_handler
    raise ex
  File "/Users/adejanovski/projets/cassandra/thelastpickle/py36/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 688, in _run_job
    result = cmd_copy(params)
  File "/Users/adejanovski/projets/cassandra/thelastpickle/py36/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 325, in __call__
    return self.handler(*args, **kwargs)
  File "/Users/adejanovski/projets/cassandra/thelastpickle/py36/lib/python3.6/site-packages/azure/cli/core/__init__.py", line 784, in default_command_handler
    return op(**command_args)
  File "/Users/adejanovski/projets/cassandra/thelastpickle/py36/lib/python3.6/site-packages/azure/multiapi/storage/v2018_11_09/blob/baseblobservice.py", line 1968, in get_blob_to_path
    with open(file_path, open_mode) as stream:
IsADirectoryError: [Errno 21] Is a directory: '/tmp/medusa-download-all-tables/system_auth/roles-5bc52802de2535edaeab188eecebb090/'
To open an issue, please run: 'az feedback'
```

Apparently the destination path must contain the file name and not just the destination directory name.
With the above suggestion, integration tests are passing.

File: tests/integration/features/steps/integration_steps.py [hunk 2]
```diff
@@ -172,6 +172,19 @@ def i_am_using_storage_provider(context, storage_provider, client_encryption):
...
+            "host_file_separator": ",",
+            "bucket_name": "medusa-integration-tests",
+            "key_file": "~/medusa_azure_credentials.json",
+            "storage_provider": "azure_blobs",
+            "fqdn": "127.0.0.1",
+            "api_key_or_username": "",
+            "api_secret_or_password": "",
+            "base_path": "/tmp",
+            "concurrent_transfers": 4,
+            "prefix": storage_prefix
```
Comment (adejanovski) [inline tests/integration/features/steps/integration_steps.py:186]:
```suggestion
            "multi_part_upload_threshold": 1 * 1024,
            "prefix": storage_prefix
```

This allows to use multipart uploads for files >= 1kB, which makes integration tests use both upload methods.

**Review by adejanovski — CHANGES_REQUESTED** [review]
Hi @ilhanadiyaman,

this is an awesome addition to Medusa, thanks a lot for pushing this PR.
There are a few places where the current implementation is broken and I've added suggestions which I tested locally with the integration tests.
Once you've applied the fixes (or implemented yours), could you please:
- Add the following entry at the top of CHANGELOG.md :
```
### 0.8.0 (TBD)
- Add Azure blob storage support (@ilhanadiyaman)
```
- squash all the commits into a single one

I'll do one last check and run the integration tests before merging once you've forced push those changes.

Thanks again!

**Comment by ilhanadiyaman** [comment]
Sorry my late response @adejanovski. Thanks for your review and suggestions. Next week, I'll review all of them and send a PR as you suggested.

File: medusa/storage/azure_storage.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        )
+        return path
+
+    def upload_blobs(self, srcs, dest):
+        return medusa.storage.azure_blobs_storage.concurrent.upload_blobs(
+            self, srcs, dest, self.bucket,
+            max_workers=self.config.concurrent_transfers,
+        )
+
+    def downlod_blobs(self, srcs, dest):
```
[the remark this answers is not in the document]
Comment (ilhanadiyaman) [reply medusa/storage/azure_storage.py:62]: Silly mistake. Nice catch!

File: medusa/storage/azure_blobs_storage/concurrent.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+from libcloud.storage.types import ObjectDoesNotExistError
+from retrying import retry
+
+import medusa
+from medusa.storage.azure_blobs_storage.azcli import AzCli
+
+MAX_UP_DOWN_LOAD_RETRIES = 5
+
+# Azure doesn't provide configuration for threshold
+MULTI_PART_UPLOAD_THRESHOLD = 64 * 1024 * 1024
```
[the remark this answers is not in the document]
Comment (ilhanadiyaman) [reply medusa/storage/azure_blobs_storage/concurrent.py:32]: Alright. I'll add the threshold configuration like S3.

**Comment by ilhanadiyaman** [comment]
Fixed all linter errors. Integration tests are passing. Hopefully, we can ship this one @adejanovski.

**Comment by adejanovski** [comment]
Awesome work @ilhanadiyaman!!

Thanks a lot for adding Azure storage support to Medusa.
I merged your changes from the command line as I needed to fix three minor issues (here's [the commit](https://github.com/thelastpickle/cassandra-medusa/commit/983234acc1b4e6bf50ff459b61ca7e667ceaef69)):

- Rebase on top of `master`
- Update setup.py to allow installing Azure optional dependencies: `sudo pip install cassandra-medusa[AZURE]`
- Remove the unused threshold constant in `storage/azure_blobs_storage/concurrent.py`

Thanks again!

**Pull request closed without merging**